### PR TITLE
8809 libzpool should leverage work done in libfakekernel

### PR DESCRIPTION
**Pull request by andy-js**
This changeset follows on from #451 which implement parallel mounting of zfs filesystems.  Additionally it attempts to cleanup libzpool by having it leverage the work done in libfakekernel.

### CODE REVIEW

**Comment by andy-js** [comment]
Not sure why this failed to build.  It's building fine on my OmniOS VM.

File: usr/src/cmd/zdb/zdb.c
Comment (prakashsurya) [inline usr/src/cmd/zdb/zdb.c]: I'm not sure if this is _correct_, but I was able to silence the lint issue by changing this line like so:
```
diff --git a/usr/src/cmd/zdb/zdb.c b/usr/src/cmd/zdb/zdb.c
index e5797ae29d..3df7180c18 100644
--- a/usr/src/cmd/zdb/zdb.c
+++ b/usr/src/cmd/zdb/zdb.c
@@ -95,7 +95,7 @@ int zfs_vdev_async_read_max_active;

 static const char cmdname[] = "zdb";
 uint8_t dump_opt[256];
-extern int aok;
+volatile int aok;

 typedef void object_viewer_t(objset_t *, uint64_t, void *data, size_t size);
```

Comment (andy-js) [reply usr/src/cmd/zdb/zdb.c]: Just out of curiosity... what was the lint error?

Comment (prakashsurya) [reply usr/src/cmd/zdb/zdb.c]:
```
==== lint warnings src ====

"/opt/jenkins/root/workspace/openzfs_openzfs_PR-494-64XNXXNYJTBMFXZULXEELHIF226JBJ6ZVXFRFO6EYEWBUT4REFUQ/usr/src/cmd/zdb/zdb.c", line 4152: warning: name used but not defined: aok in zdb.c(4152) (E_NAME_USED_NOT_DEF2)
```
you can see it in the build output for the automated build, [here](http://jenkins.open-zfs.org/blue/rest/organizations/jenkins/pipelines/openzfs/openzfs/branches/PR-494/runs/3/nodes/102/log/?start=0).

Comment (andy-js) [reply usr/src/cmd/zdb/zdb.c]: I am going to say that that's bogus.  Either that or we aren't linking in libc (which provides aok) which building the lint library, but that seems unlikely.

**Comment by prakashsurya** [comment]
Assuming this passes the automated testing, the libfakekernel changes in this look good to me.

I still think it would be beneficial to split this into two parts: 1. libfakekernel changes 2. parallel zfs mount changes. But, ultimately we want both changes to land in illumos, so I'm OK moving forward with them combined in a single commit like what's done here.

File: usr/src/cmd/zdb/zdb.c
Comment (prakashsurya) [reply usr/src/cmd/zdb/zdb.c]: lint can be a pain, but I think we'll need to workaround this even if the message is bogus. `volatile` is used instead of `extern` in the `libfakekernel` code, so I think it'd be reasonable to do that here?

**Comment by andy-js** [comment]
Yeah I think I agree.  I will split the parallel stuff out into a separate change.  To make lint happy we might need to breakout 'ifdef lint'.

**Comment by andy-js** [comment]
Where can I see what caused the build failure?

**Comment by yuripv** [comment]
@andy-js Click the details near the "continuous-integration/jenkins/pr-head — This commit cannot be built".  Looks like it's not your code problem, and rather CI one -- try force pushing the commit so that the build would be restarted.

**Comment by prakashsurya** [comment]
Sorry, yes, there some some issue with Jenkins; I've restarted the build.

**Comment by andy-js** [comment]
Thanks.  I think I see what's going on.

**Comment by andy-js** [comment]
@zettabot go

**Comment by prakashsurya** [comment]
The "zettabot go" comments no longer do anything. It will rebuild when it detects the PR code has changed; i.e. the force push you just made will kick off a new run automatically.

**Comment by andy-js** [comment]
Looks like the build failure was unrelated to my changes.  I see:
"could not determine the workspace used to perform the build"

**Comment by prakashsurya** [comment]
I'll restart the jenkins process, and then restart this build; that should allow this build to go through.

**Review by sebroy — APPROVED** [review]

**Comment by andy-js** [comment]
Looks like zloop was killed for some reason.  In the log I see "Timeout has been exceeded
Finished: ABORTED"

**Comment by prakashsurya** [comment]
That zloop failure looks similar to what happened the previous run that also failed zloop; see [here](http://jenkins.open-zfs.org/blue/organizations/jenkins/openzfs%2Fopenzfs/detail/PR-494/6/pipeline/211). There _might_ be some information that can be gathered from the zloop files, which can be downloaded [here](http://jenkins.open-zfs.org/job/openzfs/job/openzfs/job/PR-494/9/artifact/run-zloop-results.tar.xz). If there's nothing useful in there, can you try running zloop locally to see if you can trigger anything abnormal?

I have a hunch that zloop is getting "stuck", and then is killed by the Jenkins time limit after it fails to return. It's supposed to run for only 1.5 hours, and then has a 2 hour time limit enforce by Jenkins. In this case, it looks like the underlying ztest process was only supposed to run for 56 seconds:
```
11/29 19:31:01 ztest -VVVVV -m 2 -r 0 -R 1 -v 2 -a 9 -T 56 -P 23 -s 128m -f /var/tmp/test_results
```
but it seems as though it didn't return after 2 hours (hence hitting the time limit, and being killed).

**Comment by andy-js** [comment]
Thanks for your help.  I think I have pushed a fix, but CI has failed again complaining about not being able to find the workspace.

**Comment by prakashsurya** [comment]
Jenkins is being difficult, sorry for the trouble. I'll restart this build later today.

**Comment by andy-js** [comment]
The lint and test errors have now been fixed.  If I could get another review or two that'd be great.

File: usr/src/lib/libfakekernel/common/buf.c
Comment (gwr) [inline usr/src/lib/libfakekernel/common/buf.c]: Is this file sufficiently "generic" that it deserves to be part of libfakekernel?  or should it remain as part of libzpool?

File: usr/src/lib/libfakekernel/common/cred.c
Comment (gwr) [inline usr/src/lib/libfakekernel/common/cred.c]: I'm not sure we should keep the cred stuff in libfakekernel.
Maybe better to leave that to the consumers?
(libfksmbsrv, and libzpool after this work)

File: usr/src/lib/libfakekernel/common/kmem.c
Comment (gwr) [inline usr/src/lib/libfakekernel/common/kmem.c]: How about adding a strext.c file for this?
(like in uts/common/os)

It might not have been obvious, but we tried to keep the functions organized in files of similar names as the "real" versions.

File: usr/src/lib/libfakekernel/common/kmisc.c
Comment (gwr) [inline usr/src/lib/libfakekernel/common/kmisc.c]: Here too, how about a cyclic.c?

Would rather not see kmisc.c turn into more of a "catch" all...

File: usr/src/lib/libfakekernel/common/ksid.c
Comment (gwr) [inline usr/src/lib/libfakekernel/common/ksid.c]: Doesn't this clash with what libfksmbsrv is doing with creds, sids, and ksids?   I think we may need to keep this too in the consumers, rather than in libfakekernel (leaving this ignorant of creds or ksids).

File: usr/src/lib/libfakekernel/common/sys/cmn_err.h
Comment (gwr) [inline usr/src/lib/libfakekernel/common/sys/cmn_err.h]: Did you run into problems using the normal sys/cmn_err.h?
What happened?

**Review by gwr — COMMENTED** [review]
The rest looks reasonable, but let's chat about what to keep in libfakekernel vs. consumers.

File: usr/src/lib/libfakekernel/common/sys/cmn_err.h
Comment (andy-js) [reply usr/src/lib/libfakekernel/common/sys/cmn_err.h]: Some of the print functions have different prototypes (kernel versions are void, user space versions return number of characters printed) and this causes the compiler to generate an error.

File: usr/src/lib/libfakekernel/common/kmisc.c
Comment (andy-js) [reply usr/src/lib/libfakekernel/common/kmisc.c]: Sure.  This sounds good to me.

File: usr/src/lib/libfakekernel/common/buf.c
Comment (andy-js) [reply usr/src/lib/libfakekernel/common/buf.c]: I'm not sure, but if there's any question about it perhaps we don't need to move it out right now.

File: usr/src/lib/libfakekernel/common/kmem.c
Comment (andy-js) [reply usr/src/lib/libfakekernel/common/kmem.c]: Will do.

File: usr/src/lib/libfakekernel/common/ksid.c
Comment (andy-js) [reply usr/src/lib/libfakekernel/common/ksid.c]: I don't think so.  I don't see these functions defined there, anyway.  And besides, there's already some cred stuff in libfakekernel.

File: usr/src/lib/libfakekernel/common/cred.c
Comment (andy-js) [reply usr/src/lib/libfakekernel/common/cred.c]: If we could add the missing functionality and make all consumers happy I'd rather do that.

File: usr/src/lib/libfakekernel/common/ksid.c
Comment (gwr) [reply usr/src/lib/libfakekernel/common/ksid.c]: I know, but I'm starting to think that was a mistake.

In general, I'd rather "promote" things into libfakekernel only after we're
fairly sure they're generic enough to satisfy the needs of all consumers.
That's not at all obvious to me in the case of cred the stuff...

**Comment by prakashsurya** [comment]
@gwr when you have time, can you give the latest revision another look?

**Comment by prakashsurya** [comment]
@andy-js what are the next steps for this? besides code reviews, is there anything left to be done here?

**Comment by andy-js** [comment]
Gordon had some concerns about the cred stuff but I'm not so sure.  What I would like to do - if he's happy - is go ahead and revisit that in a separate change.

**Comment by prakashsurya** [comment]
I'm planning to update this to the latest master, and open the RTI after it passes the normal regression tests.

**Comment by gwr** [comment]
As we're sort of "promoting" libfakekernel to more wide-spread use, I'd really rather not have it include things that are rather SMB-specific.  I see the current cred hacking there as rather SMB-specific.
Sorry I haven't had time to offer changes to take that out, though I don't think it should be too hard.

**Comment by gwr** [comment]
That said, I hate to hold things up.  If you think the (rather limited) cred implementation is generic enough, at least for today's consumers, I guess it's OK for this to go ahead.  It does mean that if and when a consumer comes along that needs real credentials (i.e. NFS?) we'll need to change this lib and all current consumers.  If you're OK with that, then go ahead.

**Comment by prakashsurya** [comment]
@gwr While I tend to agree with you, I'd like to push this forward and get it landed as-is; since it's been lingering in the current state for a while. I'm not opposed to making the changes you suggest in a follow up patch, though.

Are you OK with that? I'm going to open the RTI as the tests have passed; feel free to chime in there if you disagree.

**Pull request closed without merging**
